### ISSUE TITLE
Improve datepicker styles

### ISSUE THREAD

**Issue by tofumatt** [WordPress/gutenberg#10181]
There's been discussion in #7621 around improving the styling/UX of the datepicker. @jasmussen has some great ideas here: https://github.com/WordPress/gutenberg/pull/7621#issuecomment-401301889 and in the same issue @karmatosed referenced a datepicker she dug: https://github.com/WordPress/gutenberg/pull/7621#issuecomment-401605022

We should do these things for sure, but I'm creating this issue separately to document the styles/UX we want to implement that we can work on after #7621 is merged.

**Comment by jasmussen** [comment]
I think this ticket is solid, but I think it's also a somewhat big undertaking. As such, I think perhaps this ticket should be in "Ideas". Because the improvements to be made involve parsing input fields like how Google Calendar does it. Not easy. Doesn't mean we shouldn't do it, but means we might need to do it in 5.1 instead.

So as to keep this repo, for now, focused on immediate future things, would you mind if I close this ticket? I've added it to Ideas for now https://github.com/WordPress/gutenberg/projects/8

**Comment by tofumatt** [comment]
Oh, hey, totally, sorry. I thought your GIF was basically it. My bad, sorry! 😅

**Comment by jasmussen** [comment]
Not at all! It's just that natural language parsing and ensuring the accessibility of it, while totally doable, is beyond my skill ;)

**Comment by tofumatt** [comment]
I don't understand why we would close an issue that isn't fixed. Closing this issue makes it appear like it's either:

* done
* not going to be addressed

but that's not the case. It also makes it difficult to search for; where do I find this issue and make suggestions or ask questions about it now that it's closed?

**Comment by sarahmonster** [comment]
It's a wee bit unclear from the description here what threshold we need to hit to consider this issue closed, but  it seems like we're looking for:

1. A simpler interface.
2. Ability to enter dates and times in a natural-language-ish kind of fashion.
3. A fully accessible solution.

I started playing around with some ideas for all three over here: https://github.com/WordPress/gutenberg/issues/12048#issuecomment-497507535; would love some additional feedback there to see if this might be an approach to closing this issue.

**Comment by sarahmonster** [comment]
Circling back here to share the suggested improvements as per #12048, let's look at the current issues with the existing datepicker:

- Entering a date using a keyboard requires three individual, disconnected selections.
- Natural-language processing may not work effectively for all languages WordPress supports.
- Closing the datepicker popover can only be done by clicking outside of the popover, causing accessibility issues.
- The boxes of different sizes everywhere creates leads to a bit of a mess. Cleaning it up declutters the interface.

Instead, let's opt for a solution that:

- Simplifies input for keyboard users.
- Displays a calendar for users who find this helpful (showing the calendar being useful to visualise dates and weekends, as well as to see when [existing posts are scheduled](https://github.com/WordPress/gutenberg/issues/13713#issuecomment-510670548)), but doesn't make calendar input the primary or most prominent entry so as to allow for a more inclusive interface.
- Reduces visual complexity and noise.
- Won't interfere with proposed changes in #470.
- Allows for more explicit "save" and "reset" behaviours. 

More discussion and mockups on #12048, but here's the solution I think would work best:
![datepicker](https://user-images.githubusercontent.com/376315/62335230-ed55f780-b4c2-11e9-8cb3-f04a67121685.png)

![datepicker-open](https://user-images.githubusercontent.com/376315/62335229-ed55f780-b4c2-11e9-8530-296cc45f323c.png)

Using a simpler input allows us to vastly simplify the interface for users who'd prefer a plaintext input as well as for users who'd prefer to use a calendar input, and it significantly reduces the cognitive overload of what was once an overly-busy and overly-complex interface.

Internationalisation is handled by hinting at the format required, combined with helpful error messaging and sensible validation (not pictured, but we can programmatically correct for user errors or format variations, as seen in the Eureka example.)

At that same time, this also solves the original problem in that there isn't a clear location in the current interface for the "Reset" button that makes it clear what that button will do, making the interface more understandable, consistent, and predictable for users.

This is a good model to follow in terms of implementation: https://dequeuniversity.com/library/aria/date-pickers/sf-date-picker

See also: http://adrianroselli.com/2019/07/maybe-you-dont-need-a-date-picker.html (Thanks @afercia!)

**Comment by afercia** [comment]
> as seen in the Eureka example

Link 🙂http://eureka2.github.io/ab-datepicker/

For the accessibility feedback, please see comment on https://github.com/WordPress/gutenberg/issues/12048#issuecomment-508932568

Note: User input validation would play a role here. Using select elements makes the task easier. For example: if the month is `2` (February) we know `30` and `31` are invalid values and `29` is only valid for leap years. Instead, parsing a string like `02/30/2020` is a bit more complicated (if it's not already handled by the adopted library).

The current Gutenberg date-picker auto-corrects invalid dates.
In the Classic Editor, submission of invalid dates is prevented:

<img width="312" alt="Screenshot 2019-08-05 at 16 12 36" src="https://user-images.githubusercontent.com/1682452/62472611-6c874c00-b79f-11e9-9c65-830ce0deb9d5.png">

**Comment by melchoyce** [comment]
It seems like Eureka handles i18n, and is accessible — could we use that?

**Comment by mapk** [comment]
It would be great to see @sarahmonster's iteration implemented if possible. Surfacing timezones as in https://github.com/WordPress/gutenberg/issues/15673 which was recently reopened is a nice addition.

**Comment by beaverbuilder** [comment]
I'm jumping into this conversation late but I've been working with `DateTimePicker` in another context and having trouble just getting it centered up and looking nice. There seem to be a lot of container divs (some don't seem like they are contributing to the layout) and several hardcoded widths. Even inside the block editor it doesn't appear to be centered and spaced properly. 

**My Question:** Is there any reason not to implement the calendar with flexbox so that it fluidly adapts to the width of its container?

<img width="407" alt="Screen Shot 2020-04-13 at 11 25 36 AM" src="https://user-images.githubusercontent.com/35972447/79138258-dbc72e00-7d79-11ea-9dd9-8b78c17ebea4.png">
DatePicker inside the Block Editor

**Comment by brentjett** [comment]
I'm jumping into this conversation late but I've been working with `DateTimePicker` in another context and having trouble just getting it centered up and looking nice. There seem to be a lot of container divs (some don't seem like they are contributing to the layout) and several hardcoded widths. Even inside the block editor it doesn't appear to be centered and spaced properly.

**My Question:** Is there any reason not to implement the calendar with flexbox so that it fluidly adapts to the width of its container?

<img width="407" alt="Screen Shot 2020-04-13 at 11 25 36 AM" src="https://user-images.githubusercontent.com/1790660/79138505-3cef0180-7d7a-11ea-9b47-a7f2471741ae.png">

**Comment by mapk** [comment]
@brentjett, I'd love to see flexbox work here! ❤️ +1 to minimizing the amount of HTML.

**Comment by youknowriad** [comment]
@jasmussen Feels like we iterated a lot on this component during the recent weeks/months. Is this done?

**Comment by jasmussen** [comment]
Yes, I think that the next actionable step requires an entirely new component, the existing one as Brent alludes to, comes with sufficient challenges, and what we have as is, works.

So I think this is closable, and any new ticket should come with a fresh appraoch and new design.
